### DISCUSSION TITLE
Using tmux/screen or rather piping its stdout/err back to sh

### DISCUSSION THREAD

**Discussion by sla-te** [amoffat/sh#703] [Q&A] [answered]
Using my solution from https://github.com/amoffat/sh/discussions/701#discussioncomment-7186549 I am now extending it to send the command via ssh to an already existing tmux session, pretty much like:

```python
with MasterSlaveSsh() as ssh:
    tmux = ssh.tmux("send-keys", "-t", "mysession_name:0")
    tmux.echo("hello", "ENTER")
```

Interestingly what arrives in the tmux session is `echohello` instead of `echo hello`.

**Comment by sla-te** [comment]
Found this: https://stackoverflow.com/questions/71306953/tmux-send-keys-swallows-spaces it seems going `tmux.echo("Space", "hello world", "Enter")` works. Feels like tmux send-keys function needs a new impl in sh?

Would go along with implementing pane piping: https://stackoverflow.com/questions/12665625/how-to-get-the-result-of-send-keys-in-tmux

**Reply by ecederstrand** [reply]
`tmux send-keys` is not implemented in sh. We just collect the arguments to any Unix binary and run the process - there's no special handling for tmux.  If tmux needs extra keywords to preserve spaces, it's not in scope of sh. Like in #701, you can create a helper function to add the spaces and newlines you need to execute your commands using tmux.

**Reply by sla-te** [reply]
Got it, I wrote a little wrapper, that will insert space between each argument and inside each argument if it finds a space inside it. What really demotivates now is, that I cant seem to find a different way to get the output back from commands I send into a tmux pane than using its `pipe-pane`. This situation results in me only getting a log file and shs really nice exception handling will not  apply anymore - would you by any chance have an idea how I could get it back in this environment?

**Reply by amoffat** [reply, marked as answer]
If I'm understanding your problem correctly, it doesn't seem related to an sh issue or an sh question, so it would be more appropriate to ask on stackoverflow than here. If you agree, please mark this discussion as closed.

**Reply by sla-te** [reply]
After more research you are probably right, I guess for using tmux in combination with ssh sh is just not the right choice, ill take a look at fabric for this part
